docs/decisions/00XX-workflow_composability_design.md

```diff
@@ -0,0 +1,245 @@
+- Type safety:
+  - `connect` enforces compatibility between source output types and target input types (or fragment contract).
+  - Allow optional `adapter` param to inject a converter executor inline if strict types differ (compatible with Option 3 registry).
+- Example:
+
+```python
+analysis = (
+    ConcurrentBuilder()
+    .participants([operation, compliance])
+    .as_connection()
+)
+
+builder = WorkflowBuilder()
+analysis_handle = builder.connect(analysis, prefix="analysis")  # merge + handle
+builder.connect(orchestrator, analysis_handle.start)
+builder.connect(analysis_handle.outputs[0], aggregator_executor)
+workflow = builder.set_start_executor(orchestrator).build()
+```
+- Pros: single workflow boundary, explicit connect vocabulary, compatibility with port semantics later. Cons: still needs ID renaming during merge and clear immutability rules for fragments.
+
+### Fluent pattern illustration with connect-only (option 2)
+
+```python
+normalize_connection = (
+    WorkflowBuilder()
+    .add_edge(Normalize(id="normalize"), Enrich(id="enrich"))
+    .set_start_executor("normalize")
+    .as_connection()
+)
+
+summarize_connection = (
+    WorkflowBuilder()
+    .add_edge(Summarize(id="summarize"), Publish(id="publish"))
+    .set_start_executor("summarize")
+    .as_connection()
+)
+
+builder = WorkflowBuilder()
+normalize_handle = builder.add_connection(normalize_connection, prefix="prep")
+summarize_handle = builder.add_connection(summarize_connection, prefix="summary")
+builder.connect(normalize_handle.output_points[0], summarize_handle.start_id)
+builder.set_start_executor(normalize_handle.start_id)
+
+workflow = builder.build()
+print("Outputs:")
+async for event in workflow.run_stream("  Hello Composition  "):
+    if isinstance(event, WorkflowOutputEvent):
+        print(event.data)
+```
+
+## Option 3: Typed Adapters and Contracts
+- Provide first-class adapter executors to bridge mismatched but intentionally compatible types instead of ad-hoc callbacks:
+  - `builder.add_adapter(source, target, adapter_fn)` sugar that injects a small Executor running adapter_fn; validated via is_type_compatible on adapter outputs.
+  - Offer canned adapters for common shapes: `ConversationToText`, `TextToConversation`, `MessagesToStructured[T]`, mirroring existing _InputToConversation/_ResponseToConversation patterns.
+- Expose explicit type contracts on fragments/workflows:
+  - `WorkflowContract` capturing `input_types`, `output_types`, and optional `output_semantics` (e.g., “conversation”, “agent_response”, “request_message”).
+  - Composition helpers use contracts to fail fast or select the right canned adapter.
+- Pros: predictable type-safe bridges and better error messages. Cons: adds small surface area but aligns with existing adapter executors already used inside SequentialBuilder.
+
+## Option 4: Port-Based Interfaces and Extension Points
```

Comment:
This would be a interesting feature as well, making it very explicit which input and outputs there are, also allowing `hosted workflows` to be instantiated with different endpoints for the different input ports.